lib/compiler/src/beam_doc.erl

```diff
@@ -0,0 +1,275 @@
+        DocV1#docs_v1{ format = DocFormat,
+                       anno = ModuleDocAnno,
+                       metadata = Meta,
+                       module_doc = create_module_doc(ModuleDoc),
+                       docs = Docs }
+    catch E:R:ST ->
+            erlang:raise(E, R, ST)
+    end.
+
+extract_meta(AST, MetaData) ->
+    foldl(fun ({attribute, _ModuleDocAnno, moduledoc, MetaDoc}, Meta) when is_map(MetaDoc) ->
+                  maps:merge(Meta, MetaDoc);
+              (_, Meta) -> Meta
+          end, MetaData, AST).
+
+-spec extract_moduledoc(AST :: [tuple()]) -> ModuleDoc :: {erl_anno:anno(), binary()} | false.
+extract_moduledoc([]) ->
+    {?DEFAULT_MODULE_DOC_LOC, none};
+extract_moduledoc([{attribute, ModuleDocAnno, moduledoc, false}| _AST]) ->
+    {ModuleDocAnno, hidden};
+extract_moduledoc([{attribute, ModuleDocAnno, moduledoc, ModuleDoc}| _AST]) when is_list(ModuleDoc) ->
+    {ModuleDocAnno, unicode:characters_to_binary(string:trim(ModuleDoc))};
+extract_moduledoc([_| AST]) ->
+    extract_moduledoc(AST).
+
+extract_docformat([]) ->
+    ?DEFAULT_FORMAT;
+extract_docformat([{attribute, _ModuleDocAnno, moduledoc, MetaFormat} | Ls]) when is_map(MetaFormat) ->
+    case maps:get(format, MetaFormat, not_found) of
+        not_found -> extract_docformat(Ls);
+        Format when is_list(Format) -> unicode:characters_to_binary(Format);
+        Format when is_binary(Format) -> Format
+    end;
+extract_docformat([_ | Ls]) ->
+    extract_docformat(Ls).
+
+-doc "
+Creates a map containing the module documentation.
+".
+-spec create_module_doc(ModuleDoc :: binary()) -> map().
+create_module_doc(ModuleDoc) when is_atom(ModuleDoc) ->
+    ModuleDoc;
+create_module_doc(ModuleDoc) when not is_atom(ModuleDoc) ->
+    create_module_doc(<<"en">>, ModuleDoc).
+
+-spec create_module_doc(Lang :: binary(), ModuleDoc :: binary()) -> map().
+create_module_doc(Lang, ModuleDoc) ->
+    #{Lang => ModuleDoc}.
+
+-spec new_state(Dirname :: unicode:chardata()) -> internal_docs().
+new_state(Dirname) ->
+    reset_state(#docs{cwd = Dirname}).
+
+-spec reset_state(State :: internal_docs()) -> internal_docs().
+reset_state(State) ->
+    State#docs{doc = none, meta = maps:new()}.
+
+-spec update_meta(State :: internal_docs(), Meta :: map()) -> internal_docs().
+update_meta(#docs{meta = Meta0}=State, Meta1) ->
+    State#docs{meta = maps:merge(Meta0, Meta1)}.
+
+-spec update_doc(State :: internal_docs(), Doc :: unicode:chardata()) -> internal_docs().
+update_doc(#docs{doc = Doc0}=State, Doc) ->
+    case Doc0 of
+        none ->
+            State#docs{doc = string:trim(Doc)};
+        Docs ->
+            State#docs{doc = Docs ++ "\n" ++ string:trim(Doc)}
+    end.
+
+-spec update_module(State :: internal_docs(), ModuleName :: unicode:chardata()) -> internal_docs().
+update_module(#docs{}=State, ModuleName) ->
+    State#docs{module_name = ModuleName}.
+
+extract_documentation(hidden, _AST, _State) -> [];
```

Comment:
In Elixir they still create the `docs` entries when the moduledoc is hidden and I think we should do the same. That is:

```erl
-module(t).
-moduledoc false.

-export([test/0]).

-doc "hello".
test() -> ok.
```

Should result in something like this:

```erl
#docs_v1{anno = 2,beam_language = erlang,
             format = <<"text/markdown">>,module_doc = hidden,
             metadata = #{},
             docs = [{{function,t,0},
                      4,
                      [<<"t()">>],
                      #{<<"en">> => <<"hello">>},
                      #{}}]}
```